skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFiles.kt

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.container.tools.skaffold
+
+import com.intellij.openapi.module.ModuleManager
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.project.rootManager
+import com.intellij.openapi.vfs.VirtualFile
+import org.jetbrains.yaml.YAMLFileType
+import java.io.ByteArrayInputStream
+import java.io.InputStream
+import java.util.Scanner
+
+private const val SKAFFOLD_API_HEADER = "apiVersion: skaffold/"
+
+/**
+ * Checks if a given file is a valid Skaffold configuration file based on type and API version.
+ */
+fun isSkaffoldFile(file: VirtualFile): Boolean {
+    with(file) {
+        if (!isDirectory && fileType is YAMLFileType && isValid) {
+            val inputStream: InputStream = ByteArrayInputStream(contentsToByteArray())
+            inputStream.use {
+                val scanner = Scanner(it)
+                // consider this YAML file as Skaffold when first line contains proper API version
+                return scanner.hasNextLine() && scanner.nextLine().startsWith(SKAFFOLD_API_HEADER)
+            }
+        }
+    }
+    return false
+}
+
+/**
+ * Finds all Skaffold configuration YAML files in the given project. "Excluded" directories from
+ * project modules are omitted and not searched (such as build output).
+ */
+internal fun findSkaffoldFiles(project: Project): List<VirtualFile> {
+    val results: MutableList<VirtualFile> = ArrayList()
+    val excludedRoots: MutableList<VirtualFile> = ArrayList()
+    ModuleManager.getInstance(project)
+        .modules.forEach { excludedRoots.addAll(it.rootManager.excludeRoots) }
+
+    findSkaffoldFiles(project.baseDir, excludedRoots, results)
+    return results
+}
+
+/**
+ * Recursively finds all Skaffold configuration YAML files in the given root file/directory.
+ * @param file file to start search from, directory or a file itself.
+ * @param excludedRoots file roots to exclude search on.
+ * @param results list to collect resulting Skaffold files into.
+ */
+internal fun findSkaffoldFiles(
+    file: VirtualFile,
+    excludedRoots: List<VirtualFile>,
+    results: MutableList<VirtualFile>
+) {
+    if (file in excludedRoots) return
+
+    if (isSkaffoldFile(file)) {
+        results.add(file)
+        return
+    }
+
+    if (file.isDirectory) {
+        file.children.forEach { findSkaffoldFiles(it, excludedRoots, results) }
```

Comment:
Instead of manually recursing through the file tree, could you use the IJ file index?

see https://github.com/GoogleCloudPlatform/google-cloud-intellij/blob/master/app-engine/java/src/com/google/cloud/tools/intellij/appengine/java/project/DefaultAppEngineAssetProvider.java#L64

Reply:
+1.  This doesn't seem to be the optimal way to find files in an IDE.

Reply:
I'll take a look at that, but the file name maybe in an arbitrary form. I'll check if it has an ability to check for a specific content and yaml file.

Reply:
More on that: Skaffold file contents may not be indexed by IDE at all and not be a part of source file sets. Something to think about while this approach will find any Skaffold files in the project base directory. Checking file indexing APIs.

Reply:
one thing to check:

I recall that the framework support infrastructure which we used for our app engine stuff and is based on the presence of a file (`appengine-web.xml`) actually looks into the contents of the file and not just the file name. You can check to see what the logic is doing there.

Reply:
Changed to use File type based index so we won't depend on file name. For flexible detection, we use similar approach, actually reading in file content for each matching file, but with framework detection.